File: KantaroBLL/Note/Nota.vb
<DebuggerDisplay("{GetText}")> Public Class Nota
    Public Valore As NotaEnum

    Public Sub New(valore As NotaEnum)
        If valore < NotaEnum.[DO] OrElse valore > NotaEnum.SI Then Throw New IndexOutOfRangeException()
        Me.Valore = valore
    End Sub

    Public ReadOnly Property IsNaturale As Boolean
        Get
            Return {NotaEnum.DO, NotaEnum.RE, NotaEnum.MI, NotaEnum.FA, NotaEnum.SOL, NotaEnum.LA, NotaEnum.SI}.Contains(Me.Valore)
        End Get
    End Property

#Region " Istanze statiche di tutte le note "
    Public Shared ReadOnly Property [DO] As Nota
        Get
            Return New Nota(NotaEnum.[DO])
        End Get
    End Property
    Public Shared ReadOnly Property DOdiesis As Nota
        Get
            Return New Nota(NotaEnum.DOdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property REb As Nota
        Get
            Return New Nota(NotaEnum.DOdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property RE As Nota
        Get
            Return New Nota(NotaEnum.RE)
        End Get
    End Property
    Public Shared ReadOnly Property REdiesis As Nota
        Get
            Return New Nota(NotaEnum.REdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property MIb As Nota
        Get
            Return New Nota(NotaEnum.REdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property MI As Nota
        Get
            Return New Nota(NotaEnum.MI)
        End Get
    End Property
    Public Shared ReadOnly Property FA As Nota
        Get
            Return New Nota(NotaEnum.FA)
        End Get
    End Property
    Public Shared ReadOnly Property FAdiesis As Nota
        Get
            Return New Nota(NotaEnum.FAdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property SOLb As Nota
        Get
            Return New Nota(NotaEnum.FAdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property SOL As Nota
        Get
            Return New Nota(NotaEnum.SOL)
        End Get
    End Property
    Public Shared ReadOnly Property SOLdiesis As Nota
        Get
            Return New Nota(NotaEnum.SOLdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property LAb As Nota
        Get
            Return New Nota(NotaEnum.SOLdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property LA As Nota
        Get
            Return New Nota(NotaEnum.LA)
        End Get
    End Property
    Public Shared ReadOnly Property LAdiesis As Nota
        Get
            Return New Nota(NotaEnum.LAdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property SIb As Nota
        Get
            Return New Nota(NotaEnum.LAdiesis)
        End Get
    End Property
    Public Shared ReadOnly Property SI As Nota
        Get
            Return New Nota(NotaEnum.SI)
        End Get
    End Property
#End Region

#Region " Definizione degli Operatori "
    Public Shared Operator +(ByVal nota As Nota, ByVal distanza As Distanza) As Nota
        If distanza.Valore < 0 Then Return nota - New Distanza() With {.Valore = -1 * distanza.Valore}
        Dim n = (nota.Valore + distanza.Semitoni) Mod 12 ' le note vanno da 0 ad 11
        Return New Nota(n)
    End Operator
    Public Shared Operator -(ByVal nota As Nota, ByVal distanza As Distanza) As Nota
        Dim n = (nota.Valore + 12 - distanza.Semitoni) Mod 12 ' le note vanno da 0 ad 11
        Return New Nota(n)
    End Operator

    Public Shared Operator =(ByVal class1 As Nota, ByVal class2 As Nota) As Boolean
        If class1 Is Nothing Then Return False
        Return class1.Equals(class2)
    End Operator

    Public Shared Operator <>(ByVal class1 As Nota, ByVal class2 As Nota) As Boolean
        Return Not class1.Equals(class2)
    End Operator
#End Region

    Private ReadOnly Property GetText() As String
        Get
            Return Me.Valore.ToString()
        End Get
    End Property

    Public Overrides Function Equals(obj As Object) As Boolean
        If TypeOf obj IsNot Nota Then Return False
        Dim n = DirectCast(obj, Nota)
        Return n.Valore.Equals(Valore)
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If TypeOf obj1 IsNot Nota Then Return False
        Return DirectCast(obj1, Nota).Equals(obj2)
    End Function

    Public Overrides Function ToString() As String
        Return Me.ToString(False)
    End Function

    Public Overloads Function ToString(bemollePrefer As Boolean) As String
        Dim sb As New Text.StringBuilder
        If IsNaturale Then
            Dim txtNota = Me.Valore.ToString

            sb.Append(txtNota.Substring(0, 1).ToUpper)
            sb.Append(txtNota.Substring(1).ToLower)
        Else
            If bemollePrefer Then
                Dim succ = Me + Semitono.Value
                sb.Append(succ.ToString)
                sb.Append("b")
            Else
                Dim prec = Me - Semitono.Value
                sb.Append(prec.ToString)
                sb.Append("#")
            End If
        End If
        Return sb.ToString()
    End Function

    Public Shared Function TryParse(value As String, ByRef result As Nota) As Boolean
        Dim n As NotaEnum
        Dim testo As String
        If value.EndsWith("b") Then
            If Not [Enum].TryParse(value.Substring(0, value.Length - 1).ToUpper, n) Then Return False
            testo = String.Format("{0}", New Nota(n) - 1 / 2).ToUpper
        Else
            testo = value.ToUpper
        End If
        If Not [Enum].TryParse(testo.Replace("#", "diesis"), n) Then Return False
        result = New Nota(n)
        Return True
    End Function

End Class

File: Enzo.Music/Note/Nota.cs
using System.Diagnostics;
using System.Numerics;
using static System.Net.Mime.MediaTypeNames;

namespace Enzo.Music;

[DebuggerDisplay("{GetText}")]
public class Nota : IAdditionOperators<Nota, Distanza, Nota>, IAdditionOperators<Nota, float, Nota>
    , ISubtractionOperators<Nota, Distanza, Nota>, ISubtractionOperators<Nota, float, Nota>
    , IEqualityOperators<Nota, Nota, bool>
{
    public NotaEnum Valore;

    public Nota(NotaEnum value)
    {
        if (value < NotaEnum.DO || value > NotaEnum.SI) throw new IndexOutOfRangeException();
        Valore = value;
    }

    public bool IsNaturale => new NotaEnum[] { NotaEnum.DO, NotaEnum.RE, NotaEnum.MI, NotaEnum.FA, NotaEnum.SOL, NotaEnum.LA, NotaEnum.SI }.Contains(Valore);

    #region Istanze statiche di tutte le note
    public static Nota DO => new(NotaEnum.DO);
    public static Nota DOdiesis => new(NotaEnum.DOdiesis);
    public static Nota REb => new(NotaEnum.DOdiesis);
    public static Nota RE => new(NotaEnum.RE);
    public static Nota REdiesis => new(NotaEnum.REdiesis);
    public static Nota MIb => new(NotaEnum.REdiesis);
    public static Nota MI => new(NotaEnum.MI);
    public static Nota FA => new(NotaEnum.FA);
    public static Nota FAdiesis => new(NotaEnum.FAdiesis);
    public static Nota SOLb => new(NotaEnum.FAdiesis);
    public static Nota SOL => new(NotaEnum.SOL);
    public static Nota SOLdiesis => new(NotaEnum.SOLdiesis);
    public static Nota LAb => new(NotaEnum.SOLdiesis);
    public static Nota LA => new(NotaEnum.LA);
    public static Nota LAdiesis => new(NotaEnum.LAdiesis);
    public static Nota SIb => new(NotaEnum.LAdiesis);
    public static Nota SI => new(NotaEnum.SI);
    #endregion

    #region Definizione degli Operatori

    public static Nota operator +(Nota left, Distanza right)
    {
        if (right.Valore < 0) return left - new Distanza { Valore = -1 * right.Valore };
        NotaEnum n = (NotaEnum)(((int)left.Valore + right.Semitoni) % 12); // le note vanno da 0 ad 11
        return new Nota(n);
    }

    public static Nota operator +(Nota left, float right)
    {
        return left + (Distanza)right;
    }

    public static Nota operator -(Nota left, Distanza right)
    {
        NotaEnum n = (NotaEnum)(((int)left.Valore + 12 - right.Semitoni) % 12); // le note vanno da 0 ad 11
        return new Nota(n);
    }

    public static Nota operator -(Nota left, float right)
    {
        return left - (Distanza)right;
    }

    public static bool operator ==(Nota? left, Nota? right)
    {
        if (left is null) return false;
        return left.Equals(right);
    }

    public static bool operator !=(Nota? left, Nota? right)
    {
        if (left is null) return false;
        return !left.Equals(right);
    }
    #endregion

    private string GetText => Valore.ToString();

    public override bool Equals(object? obj)
    {
        if (obj == null || obj is not Nota) return false;
        return Valore.Equals(((Nota?)obj)?.Valore);
    }

    public static bool Equals(Nota? obj1, object? obj2)
    {
        if (obj1 == null || obj2 == null) return false;
        return obj1.Equals(obj2);
    }

    public override string ToString()
    {
        return ToString(false);
    }

    public string ToString(bool bemollePrefer)
    {
        var sb = new System.Text.StringBuilder();
        if (IsNaturale)
        {
            var txtNota = Valore.ToString();
            sb.Append(txtNota[..1].ToUpper());
            sb.Append(txtNota[1..].ToLower());
        }
        else
        {
            if (bemollePrefer)
            {
                Nota succ = this + Semitono.Value;
                sb.Append(succ.ToString()).Append('b');
            }
            else
            {
                Nota prec = this - Semitono.Value;
                sb.Append(prec.ToString()).Append('#');
            }
        }
        return sb.ToString();
    }

    public override int GetHashCode()
    {
        return Valore.GetHashCode();
    }

    public static bool TryParse(string value, out Nota? result)
    {
        NotaEnum n;
        string testo;
        if (value.EndsWith("b")) {
            if (!Enum.TryParse(value[..^1].ToUpper(), out n))
            {
                result = null;
                return false;
            }

            testo = string.Format("{0}", new Nota(n) - .5F).ToUpper();
        } else {
            testo = value.ToUpper();
        }
        if (!Enum.TryParse(testo.Replace("#", "diesis"), out n))
        {
            result = null;
            return false;
        }

        result = new Nota(n);
        return true;
    }
}
